### PR TITLE
Optix 7.2 update

### PR DESCRIPTION
**Pull request by Hazardu**
Upgrade to cmake 7.2
Did not work with cuda 11.0.2, updating to 11.1.0 solved the problem of missing ptrs to optix functions

### CODE REVIEW

File: src/irr/video/CCUDAHandler.h [hunk 1]
```diff
@@ -245,6 +245,14 @@ class CCUDAHandler
...
 				//
 				auto* getObject() const {return obj.get();}
+				void setObject(core::smart_refctd_ptr<ObjType>&& otherObj)
+				{
+					obj = std::move(otherObj);
+				}
+				auto* getCudaResourceHandle() const
+				{
+					return &cudaHandle;
+				}
```
Comment (devshgraphicsprogramming) [inline src/irr/video/CCUDAHandler.h]: dont add these

File: examples_tests/37.OptiXTriangle/main.cpp [hunk 2]
```diff
@@ -213,7 +213,6 @@ int main()
 	OptixPipelineLinkOptions pipeline_link_options = {};
 	pipeline_link_options.maxTraceDepth = 5;
 	pipeline_link_options.debugLevel = OPTIX_COMPILE_DEBUG_LEVEL_FULL;
-	pipeline_link_options.overrideUsesMotionBlur = false;
```
Comment (devshgraphicsprogramming) [inline examples_tests/37.OptiXTriangle/main.cpp:216]: what happened to `overrideUsesMotionBlur ` in the newer SDK?

Comment (Hazardu) [reply examples_tests/37.OptiXTriangle/main.cpp:216]: its gone

File: examples_tests/37.OptiXTriangle/main.cpp [hunk 3]
```diff
@@ -244,14 +243,14 @@ int main()
 
 		auto createRecord = [&](cuda::CCUDAHandler::GraphicsAPIObjLink<video::IGPUBuffer>* outLink, const auto& recordData) -> CUdeviceptr
 		{
-			outLink->obj = core::smart_refctd_ptr<video::IGPUBuffer>(driver->createFilledDeviceLocalGPUBufferOnDedMem(sizeof(recordData),&recordData), core::dont_grab);
```
Comment (devshgraphicsprogramming) [inline examples_tests/37.OptiXTriangle/main.cpp:247]: undo everything below this line, I need to redo it anyway

File: include/irr/ext/OptiX/IDenoiser.h [hunk 4]
```diff
@@ -66,6 +74,16 @@ class IDenoiser final : public core::IReferenceCounted
 										inputOffsetX,inputOffsetY,outputLayer,
 										scratchBuffer.asBuffer.pointer+scratchBufferOffset,scratchSizeInBytes);
 		}
+		void createTilesForDenoising(
+			void* inputBuffer,
+			void* outputBuffer,
```
Comment (devshgraphicsprogramming) [inline include/irr/ext/OptiX/IDenoiser.h]: change these two pointers into `CUdeviceptr`

File: examples_tests/39.DenoiserTonemapper/main.cpp [hunk 5]
```diff
@@ -45,10 +45,12 @@ bool check_error(bool cond, const char* message)
 		os::Printer::log(message, ELL_ERROR);
 	return cond;
 }
-/*
-	if (check_error(,"!"))
-		return error_code;
-*/
+
+constexpr uint32_t overlap = 64;
+constexpr uint32_t tileWidth = 1920, tileHeight = 1080; //or 1920x1080
```
Comment (devshgraphicsprogramming) [inline examples_tests/39.DenoiserTonemapper/main.cpp]: let me know how much memory the denoiser says it needs for the scratch size with a 1080p tile

Comment (Hazardu) [reply examples_tests/39.DenoiserTonemapper/main.cpp]: 470 081 920 bytes which is 448 MiB

Comment (devshgraphicsprogramming) [reply examples_tests/39.DenoiserTonemapper/main.cpp]: Try using 1024x1024 a 2048x2048 tile, it might expose logic errors when your test image is not divisible by tile size exactly

File: examples_tests/39.DenoiserTonemapper/main.cpp [hunk 6]
```diff
@@ -61,10 +63,10 @@ int main(int argc, char* argv[])
 	params.Vsync = true;
 	params.Doublebuffer = true;
 	params.Stencilbuffer = false;
-	params.StreamingDownloadBufferSize = 256*1024*1024; // change in Vulkan fo
+	params.StreamingDownloadBufferSize = 256 *4 * 1024 * 1024; // change in Vulkan fo
```
Comment (devshgraphicsprogramming) [inline examples_tests/39.DenoiserTonemapper/main.cpp]: remember to change back and do download in 64mb sized parts

File: examples_tests/39.DenoiserTonemapper/main.cpp [hunk 7]
```diff
@@ -454,31 +456,31 @@ void main()
...
+						{
+							const auto& assetMetaChannelName = exrmeta->getName();
+							auto found = assetMetaChannelName.find(channelName.value());
+							if (found >= firstChannelNameOccurence)
+								continue;
+							firstChannelNameOccurence = found;
+							pickedChannel = std::distance(contents.begin(), &asset);
+						}
 					}
-				}
```
Comment (devshgraphicsprogramming) [inline examples_tests/39.DenoiserTonemapper/main.cpp:474]: what happened here?

File: examples_tests/39.DenoiserTonemapper/main.cpp [hunk 8]
```diff
@@ -586,152 +584,159 @@ void main()
...
 		// upload image channels and register their buffer
 		{
 			asset::ICPUBuffer* buffersToUpload[EII_COUNT];
-			for (uint32_t j=0u; j<denoiserInputCount; j++)
+			for (uint32_t j = 0u; j < denoiserInputCount; j++)
 				buffersToUpload[j] = param.image[j]->getBuffer();
-			auto gpubuffers = driver->getGPUObjectsFromAssets(buffersToUpload,buffersToUpload+denoiserInputCount,&assetConverter);
+			auto gpubuffers = driver->getGPUObjectsFromAssets(buffersToUpload, buffersToUpload + denoiserInputCount, &assetConverter);
 
 			imagePixelBuffer = core::smart_refctd_ptr<IGPUBuffer>(gpubuffers->operator[](0)->getBuffer());
```
Comment (Hazardu) [inline examples_tests/39.DenoiserTonemapper/main.cpp:722]: after changing tile size to 1024 by 1024, images are skipped
`
While executing a kernel, the device encountered a load or store instruction on an invalid memory address. This leaves the process in an inconsistent state and any further CUDA work will return the same error. To continue using CUDA, the process must be terminated and relaunched.
`

File: examples_tests/39.DenoiserTonemapper/main.cpp [hunk 9]
```diff
@@ -536,22 +538,18 @@ void main()
...
-					{
-						os::Printer::log(imageIDString + "Image too large to download from GPU in one piece!", ELL_ERROR);
-						outParam = {};
-						continue;
-					}
+					uint32_t bytesize = extent.height * region.bufferRowLength * outParam.colorTexelSize;
+					if (bytesize > params.StreamingDownloadBufferSize || extent.width > tileWidth || extent.height > tileHeight)	//instead of refusing to denoise image, tile it
+						tileImageCheck[i] = true;
+					else
+						tileImageCheck[i] = false;
```
Comment (devshgraphicsprogramming) [inline examples_tests/39.DenoiserTonemapper/main.cpp]: always tile your image

File: examples_tests/39.DenoiserTonemapper/main.cpp [hunk 10]
```diff
@@ -586,152 +584,159 @@ void main()
...
-			maxResolution[outParam.denoiserType][0] = core::max(maxResolution[outParam.denoiserType][0],outParam.width);
-			maxResolution[outParam.denoiserType][1] = core::max(maxResolution[outParam.denoiserType][1],outParam.height);
 		}
 	}
+	for (size_t i = 0; i < inputFilesAmount; i++)
+	{
+		auto& outParam = images[i];
+		maxResolution[outParam.denoiserType][0] = core::min(maxResolution[outParam.denoiserType][0], tileWidthWithOverlap)*2;
+		maxResolution[outParam.denoiserType][1] = core::min(maxResolution[outParam.denoiserType][1], tileHeightWithOverlap)*2;
+	}
```
Comment (devshgraphicsprogramming) [inline examples_tests/39.DenoiserTonemapper/main.cpp]: First of all, just iterate over denoiser types and not input images.

Secondly WHY ARE YOU MULTIPLYING by 2 ?

Comment (devshgraphicsprogramming) [reply examples_tests/39.DenoiserTonemapper/main.cpp]: actually it would be good to not clamp `maxResolution` here and use the tile size directly

File: examples_tests/39.DenoiserTonemapper/main.cpp [hunk 11]
```diff
@@ -586,152 +584,159 @@ void main()
...
+			if (maxResolution[i][0] == 0u || maxResolution[i][1] == 0u)
 			{
 				denoiser = nullptr;
 				continue;
 			}
-
+				
 			OptixDenoiserSizes m_denoiserMemReqs;
-			if (denoiser->computeMemoryResources(&m_denoiserMemReqs, maxResolution[i])!=OPTIX_SUCCESS)
+			if (denoiser->computeMemoryResources(&m_denoiserMemReqs, maxResolution[i]) != OPTIX_SUCCESS)
```
Comment (devshgraphicsprogramming) [inline examples_tests/39.DenoiserTonemapper/main.cpp]: instead of clamping max resolution, just use tile sizes here

File: examples_tests/39.DenoiserTonemapper/main.cpp [hunk 12]
```diff
@@ -586,152 +584,159 @@ void main()
...
+		auto& outParam = images[i];
+		maxResolution[outParam.denoiserType][0] = core::min(maxResolution[outParam.denoiserType][0], tileWidthWithOverlap)*2;
+		maxResolution[outParam.denoiserType][1] = core::min(maxResolution[outParam.denoiserType][1], tileHeightWithOverlap)*2;
+	}
+
 
 #define DENOISER_BUFFER_COUNT 3u //had to change to a define cause lambda was complaining about it not being a constant expression when capturing
 	// keep all CUDA links in an array (less code to map/unmap
-	cuda::CCUDAHandler::GraphicsAPIObjLink<video::IGPUBuffer> bufferLinks[DENOISER_BUFFER_COUNT];
+	cuda::CCUDAHandler::GraphicsAPIObjLink<video::IGPUBuffer> bufferLinks[DENOISER_BUFFER_COUNT + 1];
```
Comment (devshgraphicsprogramming) [inline examples_tests/39.DenoiserTonemapper/main.cpp]: why did you add an extra link?

File: examples_tests/39.DenoiserTonemapper/main.cpp (hunk 10 above)
Comment (Hazardu) [reply examples_tests/39.DenoiserTonemapper/main.cpp]: well, size of scratch is currently not calculated properly, and it turns out ~1.3 times smaller than needed

File: examples_tests/39.DenoiserTonemapper/main.cpp [hunk 13]
```diff
@@ -586,152 +584,159 @@ void main()
...
 				denoiser = nullptr;
 				continue;
 			}
-
 			denoisers[i].stateOffset = denoiserStateBufferSize;
 			denoiserStateBufferSize += denoisers[i].stateSize = m_denoiserMemReqs.stateSizeInBytes;
-			scratchBufferSize = core::max(scratchBufferSize,denoisers[i].scratchSize = m_denoiserMemReqs.recommendedScratchSizeInBytes);
-			pixelBufferSize = core::max(pixelBufferSize,core::max(asset::getTexelOrBlockBytesize(EF_R32G32B32A32_SFLOAT),(i+1u)*forcedOptiXFormatPixelStride)*maxResolution[i][0]*maxResolution[i][1]);
+			scratchBufferSize = core::max(scratchBufferSize, denoisers[i].scratchSize = m_denoiserMemReqs.withOverlapScratchSizeInBytes);
+			pixelBufferSize = core::max(pixelBufferSize, core::max(asset::getTexelOrBlockBytesize(EF_R32G32B32A32_SFLOAT), (i + 1u) * forcedOptiXFormatPixelStride) * maxResolution[i][0] * maxResolution[i][1]);
```
Comment (devshgraphicsprogramming) [inline examples_tests/39.DenoiserTonemapper/main.cpp]: `pixelBufferSize` is important because it gets used for storing the actual image as a whole for autoexposure and luminancce measure, you're gonna have a problem if this buffer is sized for a tile and not a whole image.

File: examples_tests/39.DenoiserTonemapper/main.cpp [hunk 14]
```diff
@@ -773,95 +778,144 @@ void main()
...
+							denoiserOutput.pixelStrideInBytes = 0u;
+							denoiserOutput.format = forcedOptiXFormat;
+							if (denoiser.m_denoiser->invoke(m_cudaStream, &denoiserParams, denoiserInputs, denoiserInputs + denoiserInputCount, &denoiserOutput, fakeScratchLink, denoiser.scratchSize) != OPTIX_SUCCESS)
+							{
+								os::Printer::log(makeImageIDString(i) + "Could not invoke the denoiser sucessfully, skipping image!", ELL_ERROR);
+								continue;
+							}
+						}
+					}
+				}
```
Comment (devshgraphicsprogramming) [inline examples_tests/39.DenoiserTonemapper/main.cpp]: duplicate code is bad, just tile always

File: examples_tests/39.DenoiserTonemapper/main.cpp [hunk 15]
```diff
@@ -45,10 +45,13 @@ bool check_error(bool cond, const char* message)
...
-/*
-	if (check_error(,"!"))
-		return error_code;
-*/
+
+constexpr uint32_t overlap = 64;
+//constexpr uint32_t tileWidth = 1920/2, tileHeight = 1080/2;
+constexpr uint32_t tileWidth = 1024, tileHeight = 1024;
+constexpr uint32_t tileWidthWithOverlap = tileWidth + overlap * 2;
+constexpr uint32_t tileHeightWithOverlap = tileHeight + overlap * 2;
```
Comment (devshgraphicsprogramming) [inline examples_tests/39.DenoiserTonemapper/main.cpp:52]: you don't need these two variables

File: examples_tests/39.DenoiserTonemapper/main.cpp [hunk 16]
```diff
@@ -423,10 +424,9 @@ void main()
 
 		return imageIDString;
 	};
-	bool* tileImageCheck = new bool[inputFilesAmount];
 	core::vector<ImageToDenoise> images(inputFilesAmount);
 	// load images
-	uint32_t maxResolution[EII_COUNT][2] = { 0 };
+	uint32_t maxResolution[2] = { 0,0 };
```
Comment (devshgraphicsprogramming) [inline examples_tests/39.DenoiserTonemapper/main.cpp:428]: no still keep separate maxResolution per denoiser type! just keep this as it was!

File: examples_tests/39.DenoiserTonemapper/main.cpp (hunk 6 above)
Comment (devshgraphicsprogramming) [reply examples_tests/39.DenoiserTonemapper/main.cpp]: tiled download would be nice.

**Comment by Hazardu** [comment]
PR put on hiatus, after OptiX fixes their bugs tiled download will be added

**Comment by devshgraphicsprogramming** [comment]
@Hazardu post the link to the NVidia forum bugreport you made

**Comment by Hazardu** [comment]
https://forums.developer.nvidia.com/t/optix-7-2-tiled-denoiser-errors-in-output/158274
